api/app/models/project.rb

```diff
@@ -124,6 +124,7 @@ class Project < ApplicationRecord
   before_create :assign_publisher_defaults!
   after_commit :trigger_creation_event, on: [:create]
   after_commit :queue_reindex_children_job
+  before_validation :ensure_restricted_access_notice_content
```

Comment:
Minor things, but I really think having ! at the end of hooks is a helpful idiom to follow. Not saying change them where you see them in existing code, but it's a pattern we follow in newer Rails apps and any new hooks should do it.

Additionally, it isn't a problem here, but in later versions of rubocop-rails, `before_validation` hooks should come before `before_create` / `before_update` (and certainly before `after_commit`). This helps developers reading the hooks follow the logical path of execution for callbacks, so you can read something like:

```
before_validation :ensure_restricted_access_notice_content!
before_create :assign_publisher_defaults!
before_update :prepare_to_reindex_children, if: :draft_changed?
after_commit :trigger_creation_event, on: [:create]
after_commit :queue_reindex_children_job
```
... and be able to see each step more or less in the order it executes.

Reply:
In addition, it feels like we should probably add presence validation for these fields just to make sure nothing is broken:

```ruby
validates :restricted_access_heading, :restricted_access_body, presence: { if: :restricted_access? }
```

That'll go down around line 138 with the other validations.